annotated.go

```diff
@@ -89,3 +91,296 @@ func (a Annotated) String() string {
+	}
+	return nil
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	return resultTagsAnnotation{tags}
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If a single annotation appears multiple times, all annotations
+// will be ignored.
+//
+// For example,
+//
+//  fx.Provide(
+//    fx.Annotate(
+//      NewGateWay,
+//      fx.ParamTags(`name:"ro" optional:"true"`),
+//      fx.ParamTags(`name:"rw"),
+//      fx.ResultTags(`name:"foo"`)
+//    )
+//  )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+	annotations := newAnnotations(fType)
+
+	for _, ann := range anns {
+		if e := ann.apply(&annotations); e != nil {
+			return annotationError{
+				target: f,
+				err:    e,
+			}
+		}
+	}
+
+	ins := annotations.Ins
+	outs := annotations.Outs
+
+	newF := func(args []reflect.Value) []reflect.Value {
+		var fParams, fResults []reflect.Value
+		if annotations.annotatedIn {
+			fParams = make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+		} else {
+			fParams = args
+		}
+		fResults = fVal.Call(fParams)
+		if annotations.annotatedOut {
+			var errValue reflect.Value
+			returnsError := len(outs) > 1
+
+			// wrap the result in an annotated struct
+			results := reflect.New(outs[0]).Elem()
+			// aggregate the errors from invoked function
+			// into one error.
+			if returnsError {
+				errValue = reflect.New(outs[1]).Elem()
+			}
+
+			var errResults error
+			for i := 0; i < numOut; i++ {
+				if fResults[i].Type() == _typeOfError {
+					var errResult error
+					if !fResults[i].IsNil() {
+						errResult = fResults[i].Interface().(error)
+						errResults = multierr.Append(errResults, errResult)
```

Comment:
nit: This `var` declaration appears premature since it's not needed until inside the if. 

Further, you can simplify this a little by using the `foo, ok := bar.(baz)` style cast, and ignoring the `ok` value. To explain that further, there are two ways of casting:

```
x := foo.(Bar) // panics if foo is not a Bar, or if it's nil

x, ok := foo.(Bar)  // returns false and sets x to zero value if foo is not a Bar or if it's nil
```

So you can do the following when you don't care about the `ok` and the zero-value check will suffice.

```
x, _ := foo.(Bar)
```

Which shortens the above to:

```
if err, _ := fResults[i].Interface().(error); err != nil {
    errResults = multierr.Append(errResults, err)
}
```